File: src/main/java/gargoyle/netsquares/model/Player.java
package gargoyle.netsquares.model;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;

public abstract class Player {
    private final String name;
    private int score;

    protected Player(String name) {
        this.name = name;
        reset();
    }

    void addScore(int score) {
        this.score += score;
    }

    public String getName() {
        return name;
    }

    public int getScore() {
        return score;
    }

    @Override
    public int hashCode() {
        return Objects.hash(name);
    }

    @Override
    public boolean equals(@Nullable Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Player player = (Player) o;
        return Objects.equals(name, player.name);
    }

    @Override
    public String toString() {
        return String.format("%s \"%s\" (%d)", getClass().getSimpleName(), name, score);
    }

    protected abstract @Nullable Move move(@NotNull Board board, @NotNull Directions directions);

    void reset() {
        score = 0;
    }
}

File: src/main/kotlin/gargoyle/netsquares/model/Player.kt
package gargoyle.netsquares.model

abstract class Player protected constructor(val name: String) {
    var score = 0
        private set

    init {
        reset()
    }

    override fun equals(other: Any?): Boolean =
        when {
            this === other -> true
            other == null || javaClass != other.javaClass -> false
            else -> name == (other as Player).name
        }

    fun addScore(score: Int) {
        this.score += score
    }

    override fun hashCode(): Int {
        return name.hashCode()
    }

    abstract fun move(board: Board, directions: Directions?): Move?

    override fun toString(): String = String.format("${javaClass.simpleName} \"$name\" ($score)")

    fun reset() {
        score = 0
    }
}
